python/ray/data/dataset.py

```diff
@@ -5076,6 +5076,13 @@ def __setstate__(self, state):
         self._current_executor = None
 
     def __del__(self):
+        if cleanup_hook := getattr(self, "_cleanup_hook", None):
```

Comment:
Would love to see us avoid doing this cleanup hook. Anyway to avoid this?

Reply:
https://github.com/ray-project/ray/pull/41858#discussion_r1502068592